gap.go

```diff
@@ -268,26 +305,43 @@ func (buf *rawAdvertisementPayload) HasServiceUUID(uuid UUID) bool {
 }
 
 // ManufacturerData returns the manufacturer data in the advertisement payload.
-func (buf *rawAdvertisementPayload) ManufacturerData() []ManufacturerDataElement {
-	var manufacturerData []ManufacturerDataElement
-	for index := 0; index < int(buf.len)+4; index += int(buf.data[index]) + 1 {
-		fieldLength := int(buf.data[index+0])
-		if fieldLength < 3 {
-			continue
-		}
-		fieldType := buf.data[index+1]
-		if fieldType != 0xff {
-			continue
-		}
-		key := uint16(buf.data[index+2]) | uint16(buf.data[index+3])<<8
-		manufacturerData = append(manufacturerData, ManufacturerDataElement{
-			CompanyID: key,
-			Data:      buf.data[index+4 : index+fieldLength+1],
+func (buf *rawAdvertisementPayload) ManufacturerData() []AdvertismentDataElement {
+	var manufacturerData []AdvertismentDataElement
+	for _, data := range buf.findAllFields(0xFF) { 
+		manufacturerData = append(manufacturerData, AdvertismentDataElement{
+			UUID: New16BitUUID(uint16(data[0])+(uint16(data[1])<<8)),
+			Data: data[2:],
 		})
 	}
 	return manufacturerData
 }
 
+// ServiceData returns the service data in the advertisment payload
+func (buf *rawAdvertisementPayload) ServiceData() []AdvertismentDataElement {
+	var serviceData []AdvertismentDataElement
+	for _, data := range buf.findAllFields(0x16) { //16-Bit uuid
+		serviceData = append(serviceData, AdvertismentDataElement{
+			UUID: New16BitUUID(uint16(data[0])+(uint16(data[1])<<8)),
+			Data: data[2:],
+		})
+	}
+	for _, data := range buf.findAllFields(0x20) { //32-Bit uuid
+		serviceData = append(serviceData, AdvertismentDataElement{
+			UUID: New32BitUUID(uint32(data[0])+(uint32(data[1])<<8)+(uint32(data[2])<<16)+(uint32(data[3])<<24)),
+			Data: data[4:],
+		})
+	}
+	for _, data := range buf.findAllFields(0x21) { //128-Bit uuid
+		var uuidArray [16]byte
+		copy(uuidArray[:], data[:16])
+		serviceData = append(serviceData, AdvertismentDataElement{
+			UUID: NewUUID(uuidArray),
+			Data: data[16:],
+		})
+	}
+	return serviceData
```

Comment:
With a loop like the one that was previously in `ManufacturerData`, you could make a single loop that loops through all fields only once (and avoids lots of unnecessary heap allocations!).